server/src/citypay/citypay.service.ts

```diff
@@ -1,120 +1,135 @@
+        <flexField_3>XXXXXXXXX_XXXXXXX_X</flexField_3>
+        <itemCodeKey>900312</itemCodeKey>
+        <transactionCode>11112</transactionCode>
+        <lineItemExtraData>This is some extra line item data.</lineItemExtraData>
+        <paymentAmount>${curInvoice.dcp_grandtotal}</paymentAmount>
+        <quantity>1</quantity>
+        <sequenceNumber>${i}</sequenceNumber>
+        <unitPrice>${curInvoice.dcp_grandtotal}</unitPrice>
+      </retailPaymentRequestLineItems>`)
     }
-  }
 
-  // generates cookies by headless login to MS Applicant Portal
-  private async stealCookies(retries = MAX_RETRIES) {
-    const browserPromise = puppeteer.launch({ args: ['--no-sandbox'] });
+    return `<RetailPaymentRequest xmlns="${this.config.get('CITYPAY_DOMAIN')}">
+          <profileID>${this.config.get('CITYPAY_USERNAME')}</profileID>
+          <profilePassword>${this.config.get('CITYPAY_PASSWORD')}</profilePassword>
+          <displayTransactionDescription_1></displayTransactionDescription_1>
+          <displayTransactionDescription_2></displayTransactionDescription_2>
+          <displayTransactionDescription_3></displayTransactionDescription_3>
+          <paymentExtraData></paymentExtraData>
+          <postBackURL>${this.config.get('CITYPAY_POSTBACK')}</postBackURL>
+          <returnFromCartURL>${this.config.get('CITYPAY_RETURN_FROM_CART')}</returnFromCartURL>
+          <returnFromCheckoutURL>${this.config.get('CITYPAY_RETURN_FROM_CHECKOUT')}</returnFromCheckoutURL>
+          <agencyRequestID>${agencyRequestID}</agencyRequestID>
+          ${lineItems.join('')}
+        </RetailPaymentRequest>`;
+  }
 
-    try {
-      const browser = await browserPromise;
-      const page = await browser.newPage();
+  async getCartKey(packageId) {
+    const agencyRequestID = uuidv4();
 
-      await page.goto(MS_APPLICANT_PORTAL.SIGN_IN_PAGE);
+    const { records: [firstPackage] } = await this.crmService.get('dcp_packages', `
+        $filter=dcp_packageid eq ${packageId}
+        &$expand=dcp_dcp_package_dcp_projectinvoice_package(
+          $filter=statuscode eq ${DCP_PROJECTINVOICE_CODES.statuscode.APPROVED}
+        )`
+    );
 
-      // this gets dynamically inserted after a handshake event with MS
-      await page.waitForSelector('[action="/SignIn"]');
+    const requestXML = await this.createRequestXML(firstPackage, agencyRequestID);
 
-      // enter credentials
-      await page.type('#Username', this.config.get('MS_APPLICANT_PORTAL_USERNAME'));
+    const params = new url.URLSearchParams({ saleData: requestXML });
 
-      // Something changed MCS-portal side for this selector
-      // the selector changed from Password -> PasswordValue
-      // this code will try both just in case.
-      try {
-        await page.type('#Password', this.config.get('MS_APPLICANT_PORTAL_PASSWORD'));
-      } catch (e) {
-        await page.type('#PasswordValue', this.config.get('MS_APPLICANT_PORTAL_PASSWORD'));
-      }
+    try {
```

Comment:
Do you think you could break up this try/catch into 2 or 3 smaller ones? I've had trouble in the past debugging Applicant Portal because we do a bunch of stuff in one big try block and it's hard to tell what threw an exception. You could then have each of the smaller catches throw a more description error object with a `code` like we do elsewhere in the app.

Reply:
I decided to do one try/catch here because function call in this try/catch depends on the previous to succeed. If I split it up into different try/catch, then the an exception in the first function calls would not stop execution in subsequent function calls. 

The Applicant Portal troubles with descriptive error message is because I previously set up a poor design of re-wrapping exceptions in nest exceptions and not logging out the original error. 

In this case, the error should be logged with the line of code or specific function that yielded the error.

Reply:
actually now that you mention it... i should move the try even further up...